pkg/controller/servicebindingrequest/annotations.go

```diff
@@ -100,45 +100,71 @@ func updateUnstructuredObj(client dynamic.Interface, obj *unstructured.Unstructu
 	return err
 }
 
-// SetSBRAnnotations update existing annotations to include operator's. The annotations added are
+// SetSBRAnnotations set annotations to include SBR information. Return true if object modified.
+func SetSBRAnnotations(namespacedName types.NamespacedName,
+	obj *unstructured.Unstructured) bool {
+	var changed bool
+	annotations := obj.GetAnnotations()
+	if annotations == nil {
+		annotations = make(map[string]string)
+	}
+	if annotations[sbrNamespaceAnnotation] != namespacedName.Namespace {
+		annotations[sbrNamespaceAnnotation] = namespacedName.Namespace
+		changed = true
+	}
+
+	if annotations[sbrNameAnnotation] != namespacedName.Name {
+		annotations[sbrNameAnnotation] = namespacedName.Name
+		changed = true
+	}
+	obj.SetAnnotations(annotations)
+	return changed
+}
+
+// SetAndUpdateSBRAnnotations update existing annotations to include operator's. The annotations added are
 // referring to a existing SBR namespaced name.
-func SetSBRAnnotations(
+func SetAndUpdateSBRAnnotations(
 	client dynamic.Interface,
 	namespacedName types.NamespacedName,
 	objs []*unstructured.Unstructured,
 ) error {
 	for _, obj := range objs {
-		annotations := obj.GetAnnotations()
-		if annotations == nil {
-			annotations = make(map[string]string)
-		}
-
-		annotations[sbrNamespaceAnnotation] = namespacedName.Namespace
-		annotations[sbrNameAnnotation] = namespacedName.Name
-		obj.SetAnnotations(annotations)
-
-		if err := updateUnstructuredObj(client, obj); err != nil {
-			return err
+		if SetSBRAnnotations(namespacedName, obj) {
```

Comment:
I'd rather see the comparison in here using [this existing function](https://github.com/int8dev/service-binding-operator/blob/e822e2340e9cdcf8a899f691900ea1c49a1492de/pkg/controller/servicebindingrequest/binder.go#L441-L441).

```go
newObj := SetSBRAnnotations(namespacedName, obj)
ok, err := nestedMapComparison(obj, newObj, []string{"metadata", "annotations"})

```

Reply:
Thank you! I assume that the in-place map operations would be more cost-efficient. I will do the changes accordingly.

Reply:
To be clear, the problem with the original code is the change management might be much more error prone than comparing the data structures for changes, even if performance wise it might be worse.